discovery-provider/src/tasks/aggregates/index_aggregate_track.py

```diff
@@ -0,0 +1,138 @@
+import logging
+
+
+from src.tasks.aggregates import try_updating_aggregate_table, update_aggregate_table
+from src.tasks.celery_app import celery
+from src.tasks.index_aggregate_user import get_latest_blocknumber
+
+logger = logging.getLogger(__name__)
+
+AGGREGATE_TRACK = "aggregate_track"
+
+# UPDATE_AGGREGATE_TRACK_QUERY
+# Get a lower bound blocknumber to check for new save/repost activity
+# Find a subset of tracks that have changed after that blocknumber
+# For that subset of tracks calculate the play/repost total counts
+# Insert that count for new users or update an existing row
+UPDATE_AGGREGATE_TRACK_QUERY = """
+    WITH new_track AS (
+        SELECT
+            t.track_id AS track_id
+        FROM
+            tracks t
+        GROUP BY
+            t.track_id
+        UNION
+        ALL (
+            SELECT
+                r.repost_item_id AS track_id
+            FROM
+                reposts r
+            WHERE
+                r.is_current IS TRUE
+                AND r.repost_type = 'track'
+                AND r.is_delete IS FALSE
+                AND r.blocknumber > :prev_blocknumber
+            GROUP BY
+                r.repost_item_id
+        )
+        UNION
+        ALL (
+            SELECT
+                s.save_item_id AS track_id
+            FROM
+                saves s
+            WHERE
+                s.is_current IS TRUE
+                AND s.save_type = 'track'
+                AND s.is_delete IS FALSE
+                AND s.blocknumber > :prev_blocknumber
+            GROUP BY
+                s.save_item_id
+        )
+    )
+    INSERT INTO
+        aggregate_track (track_id, repost_count, save_count)
+    SELECT
+        DISTINCT(t.track_id),
+        COALESCE(track_repost.repost_count, 0) AS repost_count,
+        COALESCE(track_save.save_count, 0) AS save_count
+    FROM
+        tracks t
```

Comment:
The idea behind this was to limit the size of the original set of results before the joins.

In hindsight, it seems like this query may be more performant?

```
        (
            SELECT
                t.track_id AS track_id
            FROM
                tracks t
            WHERE
                t.is_current IS TRUE
            GROUP BY
                t.track_id
        ) AS t
```

Reply:
We're already filtering for is_current proposed on line 101 (similar to aggregate_user). That applies to the tracks table. This proposed subquery would essentially do the same thing and then wrap it again in another select. 

The group by is actually not necessary at all. What we're doing is saying is for any changed track, count up all reposts/saves. Then for any changed track we're going to associate a repost/save count to it. We don't want to include any duplicates which is why we have is_current = true and DISTINCT. 
https://stackoverflow.com/questions/581521/whats-faster-select-distinct-or-group-by-in-mysql

Reply:
```
DISTINCT creates a temporary table and uses it for storing duplicates.  
GROUP BY does the same, but sorts the distinct results afterwards.
```

That's a good piece of info regarding DISTINCT and GROUP BY. Thanks!

My understanding of LEFT OUTER JOINS is that we can restrict the number of rows from the beginning by adding this filtering on line 62 before we perform the two joins, instead of doing the filtering on a larger joined table that would include all tracks. By restricting the number of rows the beginning of the join query, would this allow for more performant filtering and computations for the latter part of the query?

Reply:
You can verify with a query analyzer but Postgres is smart enough to know that there's a WHERE and filter before the join

Reply:
This is what I got with what's currently in the PR:

* https://explain.depesz.com/s/brFI

These are the results when using what's suggested in this thread:

* https://explain.depesz.com/s/Gln5

Both tests were run using my `jc-discovery-2-2-clone` with `:prev_blocknumber` set to `0`, essentially running a full `_update_aggregate_track()`.

If you have experience analyzing queries in the above format, it would be good to get an explanation over that tomorrow please!

My basic understanding of both links is that with the PR we will filter for t.is_current on 3.1M rows as opposed to filtering on 1.7M rows with the suggestion in this thread. In both cases, the rows are reduced to 738k rows.

I also ran the 2 queries 5 times and came up with these results:

```
PR: (25.7+25.14+25.631+25.548+25.921)/5 = 25.588
Thread: (23.45+19.6+19.4+20.42+19.16)/5 = 20.406

(25.588 - 20.406) / 25.588 = 0.20251680475 = 20.3% faster by filtering out t.is_current before the LEFT OUTER JOINs
```

For a 20% bump in performance, I'd vote to restrict on t.is_current before the LEFT OUTER JOIN rather than after.

Thoughts?

Reply:
I'm confused... the PR change on average takes longer than the thread change right? Doesn't that mean keeping the filter at the end is faster?

Reply:
don't have too much context on this but i believe the question here is why is the subquery from faster than regular from with a filter? i believe select FROM get applied before filters, thats why a FROM with a filter as a subquery gets applied first plus it has to do the left outer join on the whole result set and then throw that work away

Reply:
Ok, thanks for getting eyes on this!

Yeah, that was my understanding of things as well @dmanjunath regarding a LEFT OUTER JOIN on a smaller set of results, but it seems as though the query optimizer will even things out.

Thanks for that link @isaacsolo , that makes sense! kk, I'll leave these queries as is, but +1 on looking at DISTINCT for a future PR and for reminding me of the DISTINCT vs GROUP BY optimizations!